### PR TITLE
http filter: Add overrideable function to http filter factory interface for specifying FilterDependencies

### PR DESCRIPTION
**Pull request by auni53**
Commit message:
* Add dependencies() function to NamedHttpFilterFactory
* Add unit testing

Additional description: I thought it would be useful to add the interface itself with documentation and get that out of the way, and I noted in the release notes that validation of the dependencies is WIP.
Risk Level: Low.
See design doc attached to #14470 for more details.

### CODE REVIEW

**Comment by auni53** [comment]
@htuch would it make sense to land chunk by itself? If so, could you help me find a reviewer to tag?

@AndresGuedez

File: docs/root/version_history/current.rst
Comment (htuch) [inline docs/root/version_history/current.rst:96]: I think you can skip this for now, since its very internally focused. I'd add a single release note when we have the filter chain deps validation working.

**Review by htuch — COMMENTED** [review]
@snowp since this filter chain validation work overlaps the filter match work you are involved with, would you be willing to be a non-Googler senior maintainer to shepherd these PRs?

File: include/envoy/server/filter_config.h
Comment (htuch) [inline include/envoy/server/filter_config.h:210]: Nit: probably `const&`

Comment (htuch) [reply include/envoy/server/filter_config.h:210]: Oh, I see what's going on. This is more similar I think to `createEmptyRouteConfigProto()` above. I think you should return a `unique_ptr` with the allocated dependencies in this case.

Comment (auni53) [reply include/envoy/server/filter_config.h:210]: I don't understand what the value of that would be; the dependency specification protos won't typically be very large, so passing it by copy seems ok to me. What do you think? 

@snowp in case you have any thoughts

Comment (htuch) [reply include/envoy/server/filter_config.h:210]: Right, the protos contain multiple strings and we try not copy them around in general.

File: include/envoy/server/filter_config.h
Comment (snowp) [inline include/envoy/server/filter_config.h:210]: `FilterDependencies` -> `FilterDependenciesPtr`

**Review by snowp — CHANGES_REQUESTED** [review]
This looks good to me, just one nit

**Review by htuch — COMMENTED** [review]
LGTM modulo @snowp comments, will hand it over to @snowp for merge. Thanks.

**Comment by auni53** [comment]
Flakes related to #12209

/retest

**Comment by auni53** [comment]
@snowp /retest seems to be stuck since some of the running tests are still queued, and the failure appears to be a flake.

**Review by snowp — APPROVED** [review]
LGTM! I restarted CI, let's see how it goes

**Pull request merged**
